winui3/DispatcherQueueUpdates.md

```diff
@@ -0,0 +1,209 @@
+
+All of these APIs are for WinUI3/Renuion, not being added to the OS.
+
+# Description
+
+## DispatcherQueueSynchronizationContext
+
+Use the `DispatcherQueueSynchronizationContext` to queue work to another thread, 
+which has an associated [DispatcherQueue](https://docs.microsoft.com/uwp/api/Windows.System.DispatcherQueue).
+See the base class
+[SynchronizationContext](https://docs.microsoft.com/dotnet/api/System.Threading.SynchronizationContext)
+for more information on how to use a SynchronizationContext.
+
+This class is only available to .NET5 or above applications.
+
+The Send method is not supported and throws a NotSupportedException.
+
+## DispatcherQueueController.CreateOnCurrentThread
+
+Creates a [DispatcherQueueController](https://docs.microsoft.com/uwp/api/Windows.System.DispatcherQueueController) 
+on the caller's thread. Use the created DispatcherQueueController to create and manage the 
+lifetime of a [DispatcherQueue](https://docs.microsoft.com/uwp/api/Windows.System.DispatcherQueue)
+to run queued tasks in priority order on the Dispatcher queue's thread.
+
+This method will fail with an error if the current thread already has a DispatcherQueue associated with it.
+You can use [DispatcherQueue.GetForCurrentThread](https://docs.microsoft.com/uwp/api/Windows.System.DispatcherQueue.GetForCurrentThread)
+to see if this is the case.
+
+See also [CreateDispatcherQueueController](https://docs.microsoft.com/en-us/windows/win32/api/dispatcherqueue/nf-dispatcherqueue-createdispatcherqueuecontroller)
+
+## DependencyObject.DispatcherQueue
+
+Gets the DispatcherQueue that this object is associated with. 
+The DispatcherQueue represents a facility that can access the DependencyObject 
+on the UI thread even if the code is initiated by a non-UI thread.
+
+
+# Examples
+
+## Creating a new thread controlled by a DispatcherQueue
+
+Create a new thread and run a Window on it.
+
+> Note that this sample requires WinUI3 running in a Desktop app
+
+```cs
+var controller = DispatcherQueueController.CreateOnDedicatedThread();
+
+_ = controller.DispatcherQueue.TryEnqueue(() =>
+{
+    var syncContext = new DispatcherQueueSyncronizationContext();
+    SynchronizationContext.SetSynchronizationContext(syncContext);
+
+    var window = new Window();
+    window.Content = new TextBlock() { Text = "Hello" };
+    window.Activate();
+});
+```
+
+## Creating a DispatcherQueue for the current thread
+
+This example shows a thread ensuring it has a DispatcherQueue, setting a Synchronization
+context onto the thread that uses that queue, and then goes into a message pump.
+
+> In a WinUI you typically don't need to write this code for the initial UI thread, as it's
+automatically generated.
+
+```cs
+static void Run()
+{
+    // Ensure we have a DispatcherQueue on this thread. 
+    if(DispatcherQueue.GetForCurrentThread() == null)
+    {
+        DispatcherQueueController.CreateOnCurrentThread();

```

Comment:
Does anything interesting happen when the controller is GC'd? Or is it fine to just throw it away - the thread retains a reference to its own controller, but that reference is inaccessilble so you have no way to shut down the dispatcher queue. Your only choice is to rip the rug out from under it at thread termination.

Reply:
@scott-moore-ms

Reply:
It should be fine to throw the controller away.  

By doing that, as you say, you lose the ability to fetch it again, but that is intentional in the design of splitting the controller from the queue - the entity that created the controller is the only one that has the ability to shut down the queue, whereas anyone on a thread should be able to access the queue to schedule their own work.  If after creating the controller, you decide you don't need to explicitly shut it down in the future (which is fairly common), then discarding it is totally fine.